ccmlib/common.py

```diff
@@ -65,6 +66,118 @@ class UnavailableSocketError(CCMError):
+
+    Attributes
+    ----------
+    filename: str | bytes | os.PathLike
+        Path to a lockfile - used for logging
+    file: TextIO
+        File handle which will be used to take a lock.
+    locked: bool
+        Current status of the lock.
+    """
+    filename: str | bytes | os.PathLike
+    file: TextIO
+    locked: bool
+
+    def __init__(self, filename: str | bytes | os.PathLike):
+        # We use append because:
+        # - if a file doesn't exist, we need to create it
+        # - we don't want to truncate existing file
+        # - we want RW access to file
+        # 'a+' is the only mode that satisfies all of this.
+        self.filename = filename
+        self.file = open(filename, 'a+')
+        self.locked = False
+
+    def try_lock(self, new_status: str = None) -> (bool, Optional[int]):
+        """Tries to take a lock. If it fails, it return PID of the process
+        that currently owns this lock.
+        """
+        assert not self.locked
+        try:
+            fcntl.flock(self.file, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except OSError:
+            (blocking_pid, _) = self.read_contents()
+            return (False, blocking_pid)
+        else:
+            self.locked = True
+            old_status = self.read_status()
+            self.write_status(new_status or old_status or '')
+
+            return (True, None)
+    
+    def lock(self, new_status: str = None):
+        assert not self.locked
+        
+        fcntl.flock(self.file, fcntl.LOCK_EX)
+        self.locked = True
+        old_status = self.read_status()
+        self.write_status(new_status or old_status or '')
+    
+    def unlock(self):
+        assert self.locked
+        fcntl.flock(self.file, fcntl.LOCK_UN)
+        self.locked = False
+    
+    def write_status(self, new_status: str):
+        assert self.locked
+        self.file.seek(0, 0)
+        self.file.truncate()
+        self.file.write(f'{os.getpid()};{new_status}')
+        self.file.flush()
+    
+    def read_contents(self) -> (Optional[int], Optional[str]):
+        """Reads the lockfile and returns pair 
+        (pid of owning process, last status)
+        """
+        self.file.seek(0, 0)
+        file_data = self.file.read()
+        try:
+            (blocking_pid, old_status) = file_data.split(';', 1)
+            blocking_pid = int(blocking_pid)
+            return (blocking_pid, old_status)
+        except:
+            return (None, None)
+    
+    def read_status(self) -> Optional[str]:
+        return self.read_contents()[1]
+
+    def __enter__(self):
+            (success, blocking_pid_opt) = self.try_lock()
+            if success:
+                return self
+            print(f"Another download running into '{os.path.dirname(self.filename)}', "
+                  f"by process '{blocking_pid_opt}'. Waiting for parallel downloading finished")
+
+            if not wait_for(func=lambda: self.try_lock()[0], timeout=3600):
```

Comment:
Maybe we can pick a smaller timeout

And print some explanations about what a user might need to do, to get rid of the lock, if needed ?

Reply:
What value would you propose? Current one is the same as it was before, I just copied it.